libgalois/include/katana/analytics/ClusteringImplementationBase.h

```diff
@@ -518,6 +520,150 @@ struct ClusteringImplementationBase {
+
+    std::vector<std::vector<uint32_t>> edges_id(graph.num_nodes());
+    std::vector<std::vector<std::tuple<EdgeWeightType, uint8_t>>> edges_data(
+        graph.num_nodes());
+
+    katana::do_all(
+        katana::iterate(uint64_t{0}, graph.num_nodes()),
+        [&](uint32_t c) {
+          for (auto ii = graph.edge_begin(c); ii != graph.edge_end(c); ++ii) {
+            auto dst = graph.GetEdgeDest(ii);
+
+            if (graph.template GetEdgeData<EdgeType>(ii) == 1) {
+              edges_id[c].push_back(
+                  graph.template GetData<CurrentCommunityId>(dst));
+              edges_data[c].push_back(std::make_tuple(
+                  graph.template GetEdgeData<EdgeWeight<EdgeWeightType>>(ii),
+                  1));
+            }
+          }
+        },
+        katana::steal());
+
+    uint64_t num_nodes_next = graph.num_nodes();
+
+    /* Serial loop to reduce all the edge counts */
+    std::vector<uint64_t> prefix_edges_count(graph.num_nodes());
+    katana::GAccumulator<uint64_t> num_edges_acc;
+    katana::do_all(
+        katana::iterate(uint64_t{0}, graph.num_nodes()), [&](uint64_t c) {
+          prefix_edges_count[c] = edges_id[c].size();
+          num_edges_acc += prefix_edges_count[c];
+        });
+
+    uint64_t num_edges_next = num_edges_acc.reduce();
+    for (uint64_t c = 1; c < graph.num_nodes(); ++c) {
+      prefix_edges_count[c] += prefix_edges_count[c - 1];
+    }
+
+    KATANA_LOG_DEBUG_ASSERT(
+        prefix_edges_count[graph.num_nodes() - 1] == num_edges_next);
+
+    // Remove all the existing node/edge properties
+    for (auto property : temp_node_property_names) {
+      if (auto r = pfg_mutable->RemoveNodeProperty(property); !r) {
+        return r.error();
+      }
+    }
+    for (auto property : temp_edge_property_names) {
+      if (auto r = pfg_mutable->RemoveEdgeProperty(property); !r) {
+        return r.error();
+      }
+    }
+
+    const katana::GraphTopology& topology = pfg_mutable->topology();
+
+    using Node = katana::GraphTopology::Node;
+    using Edge = katana::GraphTopology::Edge;
+
+    katana::LargeArray<Edge> out_indices_next;
+    out_indices_next.allocateInterleaved(num_nodes_next);
+    katana::LargeArray<Node> out_dests_next;
+    out_dests_next.allocateInterleaved(num_edges_next);
+
+    katana::do_all(
+        katana::iterate(uint64_t{0}, num_nodes_next),
+        [&](uint64_t i) {
+          out_indices_next[i] = topology.out_indices->Value(i);
+        },
+        katana::no_stats());
+
+    katana::do_all(
+        katana::iterate(uint64_t{0}, num_edges_next),
+        [&](uint64_t i) { out_dests_next[i] = topology.out_dests->Value(i); },
+        katana::no_stats());
+
+    auto topo_next = std::make_unique<katana::GraphTopology>(
+        std::move(out_indices_next), std::move(out_dests_next));
+
+    auto pfg_next = std::make_unique<katana::PropertyGraph>();
+    if (auto r = pfg_next->SetTopology(std::move(topo_next)); !r) {
```

Comment:
Return a new `PropertyGraph`. Don't modify the existing `PropertyGraph`.

Reply:
In this code, I am modifying the duplicated PropertyGraph. Original graph is not passed to this fucntion. thanks

Reply:
Why do you need to duplicate? Create a new graph with the required topology

Reply:
In the original louvain, the duplicate graph is the one that gets coarsened, while the original is left untouched for verification purposes later on. Here I am taking that duplicated graph and then filtering out edges from it